fish-rust/src/builtins/functions.rs

```diff
@@ -0,0 +1,413 @@
+            }
+            other => {
+                panic!("Unexpected retval from wgetopt_long: {}", other);
+            }
+        }
+    }
+
+    *optind = w.woptind;
+    STATUS_CMD_OK
+}
+
+pub fn functions(
+    parser: &mut parser_t,
+    streams: &mut io_streams_t,
+    args: &mut [&wstr],
+) -> Option<c_int> {
+    let cmd = args[0];
+
+    let mut opts = FunctionsCmdOpts::default();
+    let mut optind = 0;
+    let retval = parse_cmd_opts(&mut opts, &mut optind, args, parser, streams);
+    if retval != STATUS_CMD_OK {
+        return retval;
+    }
+    // Shadow our args with the positionals
+    let args = &args[optind..];
+
+    if opts.print_help {
+        builtin_print_error_trailer(parser, streams, cmd);
+        return STATUS_CMD_OK;
+    }
+
+    let describe = !opts.description.is_empty();
+    if [describe, opts.erase, opts.list, opts.query, opts.copy]
+        .into_iter()
+        .filter(|b| *b)
+        .count()
+        > 1
+    {
+        streams.err.append(wgettext_fmt!(BUILTIN_ERR_COMBO, cmd));
+        builtin_print_error_trailer(parser, streams, cmd);
+        return STATUS_INVALID_ARGS;
+    }
+
+    if opts.report_metadata && opts.no_metadata {
+        streams.err.append(wgettext_fmt!(BUILTIN_ERR_COMBO, cmd));
+        builtin_print_error_trailer(parser, streams, cmd);
+        return STATUS_INVALID_ARGS;
+    }
+
+    if opts.erase {
+        for arg in args.iter() {
+            function::remove(arg);
+        }
+        // Historical - this never failed?
+        return STATUS_CMD_OK;
+    }
+
+    if !opts.description.is_empty() {
+        // Not implemented
+        if args.len() != 1 {
+            streams.err.append(wgettext_fmt!(
+                "%ls: Expected exactly one function name\n",
+                cmd
+            ));
+            builtin_print_error_trailer(parser, streams, cmd);
+            return STATUS_INVALID_ARGS;
+        }
+        let current_func = args[0];
+
+        if !function::exists(current_func, parser) {
+            streams.err.append(wgettext_fmt!(
+                "%ls: Function '%ls' does not exist\n",
+                cmd,
+                current_func
+            ));
+            builtin_print_error_trailer(parser, streams, cmd);
+            return STATUS_CMD_ERROR;
+        }
+
+        function::set_desc(current_func, opts.description, parser);
+        return STATUS_CMD_OK;
+    }
+
+    if opts.report_metadata {
+        if args.len() != 1 {
+            streams.err.append(wgettext_fmt!(
+                BUILTIN_ERR_ARG_COUNT2,
+                cmd,
+                // This error is
+                // functions: --details: expected 1 arguments; got 2
+                // The "--details" was "argv[optind - 1]" in the C++
+                // which would just give the last option.
+                // This is broken because you could do `functions --details --verbose foo bar`, and it would error about "--verbose".
+                "--details",
```

Comment:
This is a bit dirty because it means we'll no longer use `-D` if that was passed, but given that it could mention other options, and otherwise this would mean reparsing the options or storing the options I felt that was an acceptable trade-off.